HW2/P5/physics.pyx

```diff
@@ -55,31 +56,76 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+    # in order to check the overlapping,
+    # we check if grids have the same index
+
+    # first convert XY1 to grid 1
+    # and calcualte the grid_size
+    x_grid1 = <int>(XY[i,0]/grid_spacing)
+    y_grid1 = <int>(XY[i,1]/grid_spacing)
+    grid_size = <int>((1.0 / grid_spacing) + 1)
+
+
+    #then we shall check any grid which is 2 away from this one, it's like 5 * 5
+    #and our grid is in the center of this grid
+    for curr_grid_x in range(x_grid1 - 2, x_grid1 + 3):
+
+        #we have to check if curr_grid_x is in range [0,1] before we move on
+        if (curr_grid_x > 0) and (curr_grid_x < grid_size):
+            #now we move y-axis to check
+            for curr_grid_y in range(y_grid1-2,y_grid1 + 3):
+
+                #we also have to check if y values still in range
+                    if (curr_grid_y > 0) and (curr_grid_y < grid_size):
+
+                        #after all conditions are satisfied
+                        #now we check if they have the same index as the original one
+                        #if they are overlapping, so they will have the different index, go into the loop
+                        #we have to check the grid index here
+
+                        ball_index = Grid[curr_grid_x,curr_grid_y]
+
+                        if(ball_index != -1 and ball_index != i):
+                            #acquire(&locks[ball_index])
+                            XY2 = &(XY[ball_index,0])
+                            V2 = &(V[ball_index,0])
+
+                            ################################
+                            #original code, ignore it :)
+                            ################################
+                            #for j in range(i + 1, count):
+                            #    XY2 = &(XY[j, 0])
+                            #    V2 = &(V[j, 0])
+
+                            if overlapping(XY1, XY2, R):
+
+                                # SUBPROBLEM 4: Add locking
+
+                                if not moving_apart(XY1, V1, XY2, V2):
+                                    collide(XY1, V1, XY2, V2)
+                                # give a slight impulse to help separate them
+                                for dim in range(2):
+                                    V2[dim] += eps * (XY2[dim] - XY1[dim])
+                            #release(&locks[ball_index])
```

Comment:
What happened here? Why were these acquires and releases commented out? Surely they were an integral part of the locking mechanism?